### PR TITLE
11.0 fix dropshipping secure merge moves wt

### PR DESCRIPTION
**Pull request by wtaferner**
**Description of the issue/feature this PR addresses:**
This PR is mainly created b/c drop shipping currently is not working properly with the sale order flexibility given by the sale module in case you use stock and purchase as well.

So, after some time of investigation and testing it in the wild for months (the decrease of quantities) I think I have found a good trade-off to fix two issues with the commits given:

- Sync stock move initial demand with a sale order line for a decrease as well
- Sync sale order lines with the corresponded purchase order line in case of drop shipping

**Current behavior before PR:**
- Decreasing is blocked as soon the sale order is in state ```sale```
- PO is doing stupid things after the SO is confirmed and quantities are wrongly changed in case of drop shipping and even before confirmation I think

**Desired behavior after PR is merged:**
- Make the use cases possible and safe. The fear that we get negative moves b/c  of incomplete merging (my big finding IMHO to consistently stop wrong behavior without stopping to early), should be tackled and therefore we can loosen the early blockers and replace them.

@sle-odoo @jco-odoo @pimodoo @amoyaux @nim-odoo 
Feel free to take some time to review my approach and think about the possibility how many companies will be relieved that they are not dependent on some almost impossible actions to be taken by their employees manually to get it right based on warnings which do not stop the person to change anything in the sale order which will result in inconsistent data or break aligned objects as a feature.

The first step is to see if and where the tests will fail which I kind of expect in case you guys made sure that your blockers will be in place to protect from any change that will compromise your current implementation ;-) 

@sle-odoo 
My move equilibrium is maybe not needed as you were kind of amused last time I suggested it and can be improved, but should help to spare the way until we try to merge the moves and raise due to a bad situation ultimately [there](https://github.com/odoo/odoo/compare/11.0...wtaferner:11.0-fix_dropshipping_secure_merge_moves-wt?expand=1#diff-53084fa8c10a4782fca4cd4bb9f5985bR563).

And yes, I know that you have already written that you are not interested, but the drop shipping is definitely a new big issue and honestly, when I was able to make it possible, it should not be too hard to fix the open use cases you mentioned last time in https://github.com/odoo/odoo/pull/21235#issuecomment-347864067 and of course the needed behavior you blocked before release due to a lack of time. 

If you really think it was a good functional decision (instead of a decision due to a lack of time and safety as a publisher), I can tell you, it was clearly a wrong and catastrophic decision (based on the experience with 30 users trying to deal with it in RL in a fast growing warehouse environment)

After all critics and thoughts, please keep in mind, I am suggesting this to contribute and help you and others, even if you may be of other opinion by now. I really hope that you can invest some time into this as I think it is worth it (you can see it based on my text written :stuck_out_tongue_winking_eye: )

info @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock_dropshipping/models/purchase.py [hunk 1]
```diff
@@ -2,6 +2,25 @@
...
+            self.ensure_one()
+            po_line = self.env['purchase.order.line'].search([
+                ('sale_line_id', '=', self.id),
+                ('order_id.state', '!=', 'cancel')
+            ])
+            if po_line:
+                self.move_ids = po_line.move_ids
+
+        res = super(SaleOrderLine, self).write(values)
+        return res
```
Comment (wtaferner) [inline addons/stock_dropshipping/models/purchase.py]: Note to myself: move this into a new file as it is related to sale

File: addons/purchase/models/purchase.py [hunk 2]
```diff
@@ -725,7 +725,16 @@ def _prepare_stock_moves(self, picking):
...
+
+        # Take care that there will be no negative initial demand
+        moves = self.move_ids
+        pending_moves = moves.filtered(lambda r: r.state not in ['done', 'cancel'])
+        done_moves = moves.filtered(lambda r: r.state == 'done') - pending_moves
+        returned_moves = done_moves.mapped('returned_move_ids')
+        delivered_moves = done_moves - returned_moves
+        move_equilibrium = sum(delivered_moves.mapped('move_line_ids').mapped('qty_done')) - sum(returned_moves.mapped('move_line_ids').mapped('qty_done')) + sum(pending_moves.mapped('product_uom_qty'))
+
+        if move_equilibrium + diff_quantity >= 0:
```
Comment (wtaferner) [inline addons/purchase/models/purchase.py]: Should be properly compared for float values...

File: addons/purchase/models/purchase.py [hunk 3]
```diff
@@ -725,7 +725,16 @@ def _prepare_stock_moves(self, picking):
             'warehouse_id': self.order_id.picking_type_id.warehouse_id.id,
         }
         diff_quantity = self.product_qty - qty
-        if float_compare(diff_quantity, 0.0,  precision_rounding=self.product_uom.rounding) > 0:
+
+        # Take care that there will be no negative initial demand
+        moves = self.move_ids
+        pending_moves = moves.filtered(lambda r: r.state not in ['done', 'cancel'])
+        done_moves = moves.filtered(lambda r: r.state == 'done') - pending_moves
```
Comment (wtaferner) [inline addons/purchase/models/purchase.py]: Removing of the pending moves is not needed as it is explicitly not possible to be part of...same for the other code on the sale order part.

File: addons/stock_dropshipping/models/purchase.py [hunk 4]
```diff
@@ -2,6 +2,25 @@
...
+from odoo.exceptions import UserError
+
+
+class SaleOrderLine(models.Model):
+    _inherit = 'sale.order.line'
+
+    @api.multi
+    def write(self, values):
+        if 'product_uom_qty' in values:
+            self.ensure_one()
```
Comment (wtaferner) [inline addons/stock_dropshipping/models/purchase.py]: Maybe iterate over the lines that bulk update of ```product_uom_qty``` is made possible even if I think this is never used in core and not very smart from a functional point of view.

File: addons/sale_stock/models/sale_order.py [hunk 5]
```diff
@@ -101,9 +101,20 @@ def create(self, values):
     def write(self, values):
         lines = False
         if 'product_uom_qty' in values:
-            precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
-            lines = self.filtered(
-                lambda r: r.state == 'sale' and float_compare(r.product_uom_qty, values['product_uom_qty'], precision_digits=precision) == -1)
+            self.ensure_one()
```
Comment (wtaferner) [inline addons/sale_stock/models/sale_order.py]: Maybe iterate over the lines that bulk update of ```product_uom_qty``` is made possible even if I think this is never used in core and not very smart from a functional point of view.

**Comment by jco-odoo** [comment]
@wtaferner Could you test if a next activity is created when you decrease the quantity on the so line?

**Comment by jco-odoo** [comment]
Apparently it should have been fixed in 11.2.

**Comment by wtaferner** [comment]
@jco-odoo 
Where should I test this? Is there a bugfix or anything similar for 11.0?
What do you mean by the next activity can fix this?

As the activity is mostly only for a single user (no one else can see the task/activity besides the creator) and it is impossible to teach the users all the important steps to do I doubt it will resolve the issue not even in 11.2 and therefore things are still not working out in RL.

It must be fully automated otherwise it is breaking a company's neck. I have orders which could eventually be fulfilled within 5 minutes after creation or any change accordingly, so besides the pain for the user who needs to adapt, it will just not help unfortunately like the internal note. 

If you need to do anything manual it is always too late and the unrelated fulfilling person has no time to deal the system as they are just doing what the system is telling them to do.

I would love to show you once where the real pain is situated in a personal talk, but unfortunately/luckily we are all very busy :-)

**Comment by wtaferner** [comment]
@jco-odoo @sle-odoo 
Now I get why you guys are so much into manual adaptations...if you do not show the detailed operations view you will be able to manually change the initial demand (after unlocking) which is not possible the other way, so I asked myself all the time, how the heck would a user change the initial demand according to the sale order line ordered qty?

So, at least this is something we need to tackle that the initial demand can be changed manually in both operational views and not only in one although I still think this is not enough ;-)

**Comment by sle-odoo** [comment]
@wtaferner I'm sorry. I did not wanted to comment on this pull request before thinking completely about the problematic flows and the solutions you proposed, and i still haven't found the time to do so. I told this to JCO and still he commented something I know won't suit you. But well.

Can you clearly describe why you think the dropshipping feature is not working properly?

I know you really want to propagate the decrease of quantity from the SO to the delivery orders and from the purchase to the receipts, but this design choice makes sense, although you really need to try saas11.2 to have a correct overview of what should have been done for v11 regarding the display of the information. Now you have next activities in almost all the blocking cases (we still miss the diminution of quantity to produce on MO, and some other things) that clearly display what documents are impacted by which change on another doucment. It's really better.

And yes, you are right, next activities are limited to a single user, and the way we chose it (responsible) is debatable and there is room for improvement, but we are aware of this and we are trying to find something to improve it for the next releases (although as i told you, at the moment we are lacking resources in the logistic team).

> IMHO there is a related field missing on the stock.move.line as currently we are supposed to write the reserved quantity (product_uom_qty)

Now i'm not sure but you shouldn't edit produt_uom_qty on the move line (see [here](https://github.com/odoo/odoo/blob/11.0/addons/stock/models/stock_move_line.py#L217)). Either i forgot about a use case or it's an error to make it editable on the view. I'll have a look tomorrow. But the idea is to edit the initial demand of the move when an exception occurs, not the reserved quantity on the move line. When doing this, you just make the move less reserved so you would a backorder wizard will be prompted and that's not the idea when, let' say, you're decreasing the ordered qties on a SO.

**Comment by wtaferner** [comment]
> Either i forgot about a use case or it's an error to make it editable on the view. I'll have a look tomorrow.

Ok, then I am not wrong about my perceptions...and now it feels better after knowing that there was something important missing in the detailed operations view. The possibility to write the initial demand which of course is not the reserved quantity of the ```stock.move.line```

About the rest, you can still think about my approach although I know that you are ahead two minor releases and worked something out. Leave me a link and I will take a look and give feedback, although as you know I am mostly busy with serving my customers in stable.

About dropshipping it is pretty easy and obvious.

- Create a SO with dropshipping lines
- Confirm it when the customer is fine with the quantities
- Now you can get the quote from the vendor
- Before the vendor confirms, the customer want to do changes on the SO
- Do the increase of quantity in the confirmed SO
- See that the PO gets wasted (issue n° 1)
- Take another route by confirming the PO and suddenly the vendor is telling you less or more is shipped for this PO (from the last time additional qty or not all quantity fitted the container)
- How do you tackle such a scenario for the customer? You can eventually raise the quantity in the sale order but nothing useful will happen to the PO...

So, a lot is wrong in the drop shipping world....

**Comment by sle-odoo** [comment]
@wtaferner the issues in the detailed operations should be fixed with a2d554c0a55eb00002a22434606ffe0e48d3782b

**Comment by wtaferner** [comment]
@sle-odoo 
Thank you. I have seen the changes and finally it starts to become probably possible to handle the missing gaps manually...the changes are helping a lot...anyway, as I was forced to move forward I will probably keep most of the users out for handling something manually for most of the use cases. But as always, make it possible to do manually, proceed to semi-automatic and finally fully automate a process and be able to step back for any reason and any time and most importantly keep it consistent :wink:

**Pull request closed without merging**